File: JavaNext/src/main/java/com/codevalue/archnext/TriviaBot.java
package com.codevalue.archnext;

import org.glassfish.jersey.jackson.internal.jackson.jaxrs.json.JacksonJsonProvider;
import org.jsoup.Jsoup;
import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
import org.telegram.telegrambots.meta.api.methods.polls.SendPoll;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.InvocationCallback;
import javax.ws.rs.core.MediaType;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.stream.Collectors;

public class TriviaBot extends TelegramLongPollingBot {
    private static final String TRIVIA_API = "https://opentdb.com/api.php";

    private interface UpdatedQuestionsHandler {
        void onUpdatedQuestionsAvailable(UserContext context, boolean isSuccess);
    }

    private static class UserContext {
        final Set<String> processedQuestions = new HashSet<>();
        final List<TriviaQuestion> questionsPool = new ArrayList<>();
    }

    private static final String COMMAND_PREFIX = "/";
    private static final String HELP_TEXT = "You can request next trivia question by sending /next or finish the session by sending /bye.";

    private final Random random = new Random();
    private final Client httpClient = ClientBuilder.newClient().register(JacksonJsonProvider.class);
    private final ConcurrentMap<Integer, UserContext> userContexts = new ConcurrentHashMap<>();

    @Override
    public String getBotUsername() {
        return BotConfig.getBotName();
    }

    @Override
    public String getBotToken() {
        return BotConfig.getBotToken();
    }

    @Override
    public void onUpdateReceived(Update update) {
        if (!update.hasMessage()) return;

        var message = update.getMessage();
        if (!message.hasText()) return;

        if (message.getText().startsWith(COMMAND_PREFIX)) {
            handleCommand(update);
        }
    }

    private void handleCommand(Update update) {
        var commandText = update.getMessage().getText();
        var response = switch (commandText) {
            case "/start" -> handleSessionStart(update);
            case "/bye" -> handleSessionEnd(update);
            case "/next" -> handleNextRequest(update);
            default -> getDefaultResponse(update, commandText);
        };

        sendResponse(response);
    }

    private void sendResponse(BotApiMethod<?> response) {
        if (response != null) {
            try {
                execute(response);
            } catch (TelegramApiException e) {
                System.out.println("Failed to send the response: " + e);
                e.printStackTrace();
            }
        }
    }

    private SendMessage handleSessionStart(Update update) {
        userContexts.putIfAbsent(update.getMessage().getFrom().getId(), new UserContext());
        return buildSendMessage(
                "Hello " + update.getMessage().getFrom().getFirstName() + "!\n" + HELP_TEXT,
                update
        );
    }

    private BotApiMethod<?> handleNextRequest(final Update update) {
        var userContext = userContexts.get(update.getMessage().getFrom().getId());
        if (userContext == null) {
            return buildSendMessage(
                    "I beg your pardon, but who are you? We haven't been introduced yet... (please /start me)",
                    update
            );
        }

        if (userContext.questionsPool.isEmpty()) {
            updateQuestions(
                    userContext,
                    (context, isSuccess) -> {
                        if (isSuccess) {
                            SendPoll poll = getNextQuestion(context, update);
                            sendResponse(poll);
                            return;
                        }

                        sendResponse(
                                buildSendMessage(
                                        "Sorry, " + update.getMessage().getFrom().getFirstName() + ", but I cannot find more questions now, please try again later.",
                                        update
                                )
                        );
                    }
            );

            return buildSendMessage(
                    update.getMessage().getFrom().getFirstName() + ", please wait till I find more questions for you...",
                    update
            );
        }

        return getNextQuestion(userContext, update);
    }

    private SendPoll getNextQuestion(UserContext userContext, Update update) {
        var poll = new SendPoll();
        poll.setChatId(update.getMessage().getChatId());
        poll.setType("quiz");

        var question = userContext.questionsPool.remove(userContext.questionsPool.size() - 1);
        userContext.processedQuestions.add(question.getQuestion());
        poll.setQuestion(Jsoup.parse(question.getQuestion()).text());

        var answers = question.getIncorrectAnswers()
                .stream()
                .map(answer -> Jsoup.parse(answer).text())
                .collect(Collectors.toList());

        var insertIndex = random.nextInt(answers.size());
        answers.add(insertIndex, Jsoup.parse(question.getCorrectAnswer()).text());
        poll.setOptions(answers);
        poll.setCorrectOptionId(insertIndex);

        return poll;
    }

    private void updateQuestions(final UserContext userContext, final UpdatedQuestionsHandler updateHandler) {
        httpClient
                .target(TRIVIA_API)
                .queryParam("amount", 10)
                .request(MediaType.APPLICATION_JSON)
                .async()
                .get(new InvocationCallback<TriviaQuestionsSet>() {
                    @Override
                    public void completed(TriviaQuestionsSet triviaQuestionsSet) {
                        if (triviaQuestionsSet.getResponseCode() != 0) {
                            System.out.println("Couldn't get new questions, response code: " + triviaQuestionsSet.getResponseCode());
                            updateHandler.onUpdatedQuestionsAvailable(userContext, false);
                            return;
                        }

                        userContext.questionsPool.addAll(
                                triviaQuestionsSet.getResults()
                                    .stream()
                                    .filter(question -> !userContext.processedQuestions.contains(question.getQuestion()))
                                    .collect(Collectors.toList())
                        );

                        if (userContext.questionsPool.size() == 0) {
                            updateHandler.onUpdatedQuestionsAvailable(userContext, false);
                            return;
                        }

                        updateHandler.onUpdatedQuestionsAvailable(userContext, true);
                    }

                    @Override
                    public void failed(Throwable throwable) {
                        System.out.println("Couldn't get new questions, error: " + throwable);
                        updateHandler.onUpdatedQuestionsAvailable(userContext, false);
                    }
                });
    }

    private SendMessage handleSessionEnd(Update update) {
        if (userContexts.remove(update.getMessage().getFrom().getId()) == null) return null;

        return buildSendMessage(
                "Good bye " + update.getMessage().getFrom().getFirstName() + ", see you next time!",
                update
        );
    }

    private SendMessage getDefaultResponse(Update update, String commandText) {
        return buildSendMessage(
                "Sorry, I don't understand this command: '" + commandText + "'\n" + HELP_TEXT,
                update
        );
    }

    private static SendMessage buildSendMessage(String messageText, Update update) {
        SendMessage message = new SendMessage();
        message.setChatId(update.getMessage().getChatId());
        message.setText(messageText);
        return message;
    }
}

File: Kotlin/src/main/kotlin/com/codevalue/archnext/TriviaBot.kt
package com.codevalue.archnext

import com.pengrad.telegrambot.TelegramBot
import com.pengrad.telegrambot.UpdatesListener
import com.pengrad.telegrambot.model.Update
import com.pengrad.telegrambot.request.BaseRequest
import com.pengrad.telegrambot.request.SendMessage
import com.pengrad.telegrambot.request.SendPoll
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import okhttp3.*
import org.jsoup.Jsoup
import java.io.IOException
import java.lang.RuntimeException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

object TriviaBot {
    private class UserContext(
            val processedQuestions: MutableSet<String> = hashSetOf(),
            val questionsPool: MutableList<TriviaQuestion> = mutableListOf()
    )

    private const val TRIVIA_API = "https://opentdb.com/api.php"
    private const val COMMAND_PREFIX = "/"
    private const val HELP_TEXT = "You can request next trivia question by sending /next or finish the session by sending /bye."

    private val httpClient = OkHttpClient()
    private val userContexts: ConcurrentMap<Int, UserContext> = ConcurrentHashMap()
    private val telegramBot: Lazy<TelegramBot> = lazy { TelegramBot(BotConfig.botToken) }
    private val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
    private val triviaQuestionsSetAdapter = moshi.adapter(TriviaQuestionsSet::class.java)

    private fun onUpdateReceived(update: Update) {
        if (update.message()?.text()?.startsWith(COMMAND_PREFIX) == true) {
            handleCommand(update)
        }
    }

    private fun buildSendMessage(messageText: String, update: Update): SendMessage =
            SendMessage(update.message().chat().id(), messageText)

    private fun handleCommand(update: Update) {
        val response = when (val commandText = update.message().text()) {
            "/start" -> handleSessionStart(update)
            "/bye" -> handleSessionEnd(update)
            "/next" -> handleNextRequest(update)
            else -> getDefaultResponse(update, commandText)
        }
        sendResponse(response)
    }

    private fun sendResponse(response: BaseRequest<*, *>?) {
        if (response == null) return

        try {
            telegramBot.value.execute(response)
        } catch (e: RuntimeException) {
            println("Failed to send the response: $e")
            e.printStackTrace()
        }
    }

    private fun handleSessionStart(update: Update): SendMessage {
        userContexts.putIfAbsent(update.message().from().id(), UserContext())
        return buildSendMessage(
                """
                    Hello ${update.message().from().firstName()}!
                    $HELP_TEXT
                    """.trimIndent(),
                update
        )
    }

    private fun handleNextRequest(update: Update): BaseRequest<*, *> {
        val userContext = userContexts[update.message().from().id()]
                ?: return buildSendMessage(
                        "I beg your pardon, but who are you? We haven't been introduced yet... (please /start me)",
                        update
                )

        if (userContext.questionsPool.isEmpty()) {
            updateQuestions(userContext) updateHandler@{ context, isSuccess ->
                if (isSuccess) {
                    val poll = getNextQuestion(context, update)
                    sendResponse(poll)
                    return@updateHandler
                }
                sendResponse(
                        buildSendMessage(
                                "Sorry, ${update.message().from().firstName()}, but I cannot find more question now, please try again later.",
                                update
                        )
                )
            }

            return buildSendMessage(
                    "${update.message().from().firstName()}, please wait till I find more questions for you...",
                    update
            )
        }

        return getNextQuestion(userContext, update)
    }

    private fun getNextQuestion(userContext: UserContext, update: Update): SendPoll {
        val nextQuestion = userContext.questionsPool.removeAt(userContext.questionsPool.lastIndex)
        userContext.processedQuestions += nextQuestion.question

        val correctAnswer = Jsoup.parse(nextQuestion.correctAnswer).text()
        val answers = nextQuestion.incorrectAnswers
                .map { Jsoup.parse(it).text() }
                .plus(correctAnswer)
                .shuffled()

        return SendPoll(
                update.message().chat().id(),
                Jsoup.parse(nextQuestion.question).text(),
                *answers.toTypedArray())
                .type("quiz")
                .correctOptionId(answers.indexOf(correctAnswer))
    }

    private fun updateQuestions(userContext: UserContext, updateHandler: (UserContext, Boolean) -> Unit) {
        val request = Request.Builder()
                .url("$TRIVIA_API?amount=10")
                .build()

        httpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                println("Couldn't get new questions, error occurred: $e")
                updateHandler(userContext, false)
            }

            override fun onResponse(call: Call, response: Response) {
                val triviaQuestionsSet = triviaQuestionsSetAdapter.fromJson(response.body()!!.source())
                if (triviaQuestionsSet!!.responseCode != 0) {
                    println("Couldn't get new questions, API response code: ${response.code()}")
                    updateHandler(userContext, false)
                    return
                }

                for (question in triviaQuestionsSet.results) {
                    if (!userContext.processedQuestions.contains(question.question)) {
                        userContext.questionsPool.add(question)
                    }
                }

                if (userContext.questionsPool.isEmpty()) {
                    updateHandler(userContext, false)
                    return
                }

                updateHandler(userContext, true)
            }
        })
    }

    private fun handleSessionEnd(update: Update): SendMessage? {
        if (userContexts.remove(update.message().from().id()) == null) return null

        return buildSendMessage(
                "Good bye ${update.message().from().firstName()}, see you next time!",
                update
        )
    }

    private fun getDefaultResponse(update: Update, commandText: String): SendMessage {
        return buildSendMessage(
                """
                    Sorry, I don't understand this command: '$commandText'
                    $HELP_TEXT
                    """.trimIndent(),
                update
        )
    }

    fun initialize() {
        telegramBot.value.setUpdatesListener { updates ->
            updates.filter { it.message() != null }.forEach { onUpdateReceived(it) }
            UpdatesListener.CONFIRMED_UPDATES_ALL
        }
    }
}